services/brig/src/Brig/API/OAuth.hs

```diff
@@ -0,0 +1,397 @@
+import Cassandra hiding (Set)
+import qualified Cassandra as C
+import Control.Error (assertMay, failWith, failWithM)
+import Control.Lens (view, (?~), (^?))
+import Control.Monad.Except
+import Crypto.JWT hiding (params, uri)
+import Data.ByteString.Conversion
+import Data.Domain
+import Data.Id
+import Data.Misc (PlainTextPassword (PlainTextPassword))
+import qualified Data.Set as Set
+import Data.String.Conversions (cs)
+import Data.Text.Ascii
+import Data.Time
+import Imports hiding (exp)
+import OpenSSL.Random (randBytes)
+import Polysemy (Member)
+import Servant hiding (Handler, Tagged)
+import Wire.API.Error
+import Wire.API.OAuth as OAuth
+import qualified Wire.API.Routes.Internal.Brig.OAuth as I
+import Wire.API.Routes.Named (Named (..))
+import Wire.API.Routes.Public.Brig.OAuth
+import Wire.Sem.Jwk
+import qualified Wire.Sem.Jwk as Jwk
+import Wire.Sem.Now (Now)
+import qualified Wire.Sem.Now as Now
+
+--------------------------------------------------------------------------------
+-- API Internal
+
+internalOauthAPI :: ServerT I.OAuthAPI (Handler r)
+internalOauthAPI =
+  Named @"create-oauth-client" createNewOAuthClient
+
+--------------------------------------------------------------------------------
+-- API Public
+
+oauthAPI :: (Member Now r, Member Jwk r) => ServerT OAuthAPI (Handler r)
+oauthAPI =
+  Named @"get-oauth-client" getOAuthClient
+    :<|> Named @"create-oauth-auth-code" createNewOAuthAuthorizationCode
+    :<|> Named @"create-oauth-access-token" createAccessTokenWith
+    :<|> Named @"revoke-oauth-refresh-token" revokeRefreshToken
+    :<|> Named @"get-oauth-applications" getOAuthApplications
+    :<|> Named @"revoke-oauth-account-access" revokeOAuthAccountAccess
+
+--------------------------------------------------------------------------------
+-- Handlers
+
+createNewOAuthClient :: NewOAuthClient -> (Handler r) OAuthClientCredentials
+createNewOAuthClient (NewOAuthClient name uri) = do
+  unlessM (Opt.setOAuthEnabled <$> view settings) $ throwStd $ errorToWai @'OAuthFeatureDisabled
+  credentials@(OAuthClientCredentials cid secret) <- OAuthClientCredentials <$> randomId <*> createSecret
+  safeSecret <- liftIO $ hashClientSecret secret
+  lift $ wrapClient $ insertOAuthClient cid name uri safeSecret
+  pure credentials
+  where
+    createSecret :: MonadIO m => m OAuthClientPlainTextSecret
+    createSecret = OAuthClientPlainTextSecret <$> rand32Bytes
+
+    hashClientSecret :: MonadIO m => OAuthClientPlainTextSecret -> m Password
+    hashClientSecret = mkSafePassword . PlainTextPassword . toText . unOAuthClientPlainTextSecret
+
+--------------------------------------------------------------------------------
+
+getOAuthClient :: UserId -> OAuthClientId -> (Handler r) (Maybe OAuthClient)
+getOAuthClient _ cid = do
+  unlessM (Opt.setOAuthEnabled <$> view settings) $ throwStd $ errorToWai @'OAuthFeatureDisabled
+  lift $ wrapClient $ lookupOauthClient cid
```

Comment:
```suggestion
guardOAuthEnabled :: (Handler r) a -> (Handler r) ((Handler r) a)
guardOAuthEnabled action = do
  unlessM (Opt.setOAuthEnabled <$> view settings) $ throwStd $ errorToWai @'OAuthFeatureDisabled
  action

getOAuthClient :: UserId -> OAuthClientId -> (Handler r) (Maybe OAuthClient)
getOAuthClient _ cid = guardOAuthEnabled $ do
  lift $ wrapClient $ lookupOauthClient cid
```

Then this should be called in all handler functions in this module.  If you really have to respond with different errors on different end-points, then make the error an argument to the guard function.

Reply:
IDK, it's one line. If we extract this, we even increase LOC. Following DRY not worth it here. But this is only my opinion, can change it if you want.